libs/execution_context/ebpf_native.c

```diff
@@ -0,0 +1,1219 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include "ebpf_core.h"
+#include "ebpf_native.h"
+#include "ebpf_object.h"
+#include "ebpf_program.h"
+#include "ebpf_protocol.h"
+#include "ebpf_handle.h"
+
+#define DEFAULT_PIN_ROOT_PATH "/ebpf/global"
+#define EBPF_MAX_PIN_PATH_LENGTH 256
+
+#ifndef GUID_NULL
+static const GUID GUID_NULL = {0, 0, 0, {0, 0, 0, 0, 0, 0, 0, 0}};
+#endif
+
+typedef uint64_t (*helper_function_address)(uint64_t r1, uint64_t r2, uint64_t r3, uint64_t r4, uint64_t r5);
+typedef void (*ebpf_free_native_t)(ebpf_native_t* native_object);
+
+static void
+_ebpf_native_unload_workitem(_In_ const void* module_id);
+
+typedef struct _ebpf_native_map
+{
+    map_entry_t* entry;
+    struct _ebpf_native_map* inner_map;
+    ebpf_handle_t handle;
+    ebpf_handle_t inner_map_handle;
+    int32_t original_fd;
+    int32_t inner_map_original_fd;
+    ebpf_utf8_string_t pin_path;
+    bool reused;
+    bool pinned;
+} ebpf_native_map_t;
+
+typedef struct _ebpf_native_program
+{
+    program_entry_t* entry;
+    ebpf_handle_t handle;
+} ebpf_native_program_t;
+
+typedef struct _ebpf_native
+{
+    GUID client_id;
+    metadata_table_t* table;
+    volatile int32_t reference_count;
```

Comment:
Why not derive from the common reference counting object?

Reply:
Discussed offline. `ebpf_core_object_t` also uses `int32_t` as reference count.